src/parseSchema.ts

```diff
@@ -0,0 +1,218 @@
+	private index: Map<string, ElementSchema> = new Map();
+	customTokens: CustomToken[];
+
+	constructor(schemaRoot: Block) {
+		if (schemaRoot.tag !== Reserved.rootTag) {
+			throw new Error(`Expected schema ${Reserved.rootTag}, got ${schemaRoot.tag}`);
+		}
+		this.customTokens = [];
+		schemaRoot.children.forEach(child => this.parseSchemaElement(child));
+	}
+
+	private parseSchemaElement(element: Block): void {
+		const name = getHeadString(element);
+		if (element.tag === SchemaTags.Block) {
+			this.index.set(name, this.parseBlockSchema(element));
+		} else if (element.tag === SchemaTags.Inline) {
+			this.index.set(name, this.parseInlineSchema(element));
+			const sugar = queryChildren(element, SchemaTags.Sugar);
+			if (sugar) {
+				const start = queryChildren(sugar, SchemaTags.Start);
+				const end = queryChildren(sugar, SchemaTags.End);
+				if (start && end) {
+					this.customTokens.push({
+						tag: name,
+						start: escapeRegExp(getHeadString(start)),
+						end: escapeRegExp(getHeadString(end))
+					});
+				}
+			}
+		}
+	}
+
+	private parseBlockSchema(element: Block): BlockSchema {
+		const raw = this.isRaw(element);
+		const defaultElem = queryChildren(element, SchemaTags.Default);
+		const contentBlock = queryChildren(element, SchemaTags.Content);
+		const content = contentBlock ? this.parseCardinalityRules(contentBlock) : new Map();
+		return {
+			content,
+			raw,
+			defaultElem: defaultElem && getHeadString(defaultElem)
+		};
+	}
+
+	private parseInlineSchema(element: Block): InlineSchema {
+		return queryAllChildren(element, SchemaTags.Arg).map(arg => {
+			const raw = this.isRaw(arg);
+			const contentBlock = queryChildren(arg, SchemaTags.Content);
+			const content = contentBlock ? this.parseCardinalityRules(contentBlock) : new Map();
+			return { raw, content };
+		});
+	}
+
+	private parseCardinalityRules(parent: Block): CardinalityRules {
+		const rules: CardinalityRules = new Map();
+		for (const rule of parent.children) {
+			const cardinality =
+				rule.tag === Reserved.defaultTag
+					? Cardinality.ZeroOrMore
+					: (rule.tag as Cardinality);
+			const name = getHeadString(rule);
+			rules.set(name, cardinality);
+		}
+		return rules;
+	}
+
+	private isRaw(block: Block): boolean {
+		return Boolean(queryChildren(block, SchemaTags.Raw));
+	}
+
+	getDefault(parentName: string): string | undefined {
+		const element = this.index.get(parentName);
+		if (element && this.isBlockSchema(element)) {
+			return element.defaultElem;
+		}
+		return undefined;
+	}
+
+	validateBlock(tree: Block): Error[] {
+		const schema = this.index.get(tree.tag);
+		if (schema === undefined) {
+			return [new Error(`Unknown tag ${tree.tag}`)];
+		}
+		if (!this.isBlockSchema(schema)) {
+			return [new Error(`Expected ${tree.tag} to be used as an inline tag`)];
+		}
+		const childTags = tree.children.map(child => child.tag);
+		const errors = this.validateCardinalityRules(schema.content, tree.tag, childTags);
```

Comment:
If we only validate cardinality for block elements, could `tagChilds` be computed directly inside `validateCardinalityRules()`?

Reply:
As discussed IRL, we may simplify cardinality rules for inline arguments later. For now, we'll leave it as is.